### PR TITLE
Process files only supported by plugins

### LINKED ISSUE #15079: [v3] plugins declared in `.prettierrc` are not loaded on directory
**Environments:**

- Prettier Version: 3.0.0
- Running Prettier via:  CLI
- Runtime:  v18.16.0
- Operating System:  Linux
- Prettier plugins (if any):  prettier-plugin-sh

**Steps to reproduce:**

https://github.com/un-ts/prettier/issues/273#issuecomment-1629109603

**Expected behavior:**

Like prettier v2, `.sh` files should be checked

**Actual behavior:**

`.sh` files are not checked

---

After debugging, `getSupportedFilesGlob` here https://github.com/prettier/prettier/blob/79d8a3983842d1d8a077f177ed8b983c1c596088/src/cli/expand-patterns.js#L130-L145
does not resolve `.sh` files correctly.

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Implementation of https://github.com/prettier/prettier/pull/15155#issuecomment-1723654981

Closes #15155
Fixes #15079

This PR is based on #15155 and inherits the test files. The implementation has been completely changed based on the review from @fisker.

This updates the snapshots of some existing tests. I think this is fine, what do you think?

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
I'll take a deeper look later, since I'm very busy these days, it may take a while.
Ping @thorn0, tell us what do you think if you have time, since you implement the directory expand logic at first.

File: changelog_unreleased/cli/15433.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#### Load files only supported by plugins (#15433 by @sosukesuzuki)
```
Comment (fisker) [inline changelog_unreleased/cli/15433.md:1]: How about "Process files ..."?

File: src/cli/expand-patterns.js [hunk 2]
```diff
@@ -4,6 +4,10 @@ import { fastGlob } from "./prettier-internal.js";
 
 /** @typedef {import('./context').Context} Context */
 
+function isError(pathOrError) {
+  return Object.hasOwn(pathOrError, "error");
+}
```
Comment (fisker) [inline src/cli/expand-patterns.js:9]: It should be safe to check `.error` directly.

File: src/cli/format.js [hunk 3]
```diff
@@ -22,17 +22,16 @@ class DebugError extends Error {
   name = "DebugError";
 }
 
-function handleError(context, filename, error, printedFilename) {
+function handleError(context, filename, error, printedFilename, ignoreUnknown) {
+  ignoreUnknown = context.argv.ignoreUnknown || ignoreUnknown;
```
Comment (fisker) [inline src/cli/format.js:26]: This can be replaced this a logical `||=`.

**Comment by fisker** [comment]
I understand that cases mentioned in https://github.com/prettier/prettier/pull/15155#discussion_r1328782334 can be little hard to setup and test, but can you try?

File: src/cli/expand-patterns.js (hunk 2 above)
Comment (fisker) [reply src/cli/expand-patterns.js:9]: I didn't mean to use `in`, I mean it can be removed. Do we still need it?

Comment (sosukesuzuki) [reply src/cli/expand-patterns.js:9]: no, I just misunderstood your comment.

**Comment by sosukesuzuki** [comment]
I've added tests for https://github.com/prettier/prettier/pull/15155#discussion_r1328788357 and https://github.com/prettier/prettier/pull/15155#discussion_r1328788357.

I'll add more tests for https://github.com/prettier/prettier/pull/15155#discussion_r1328796741 and https://github.com/prettier/prettier/pull/15155#discussion_r1328803145 (?) tomorrow.

**Review by fisker — APPROVED** [review]
LGTM, more tests are good, but totally optional.

**Comment by fisker** [comment]
Lint is failing.

File: src/cli/format.js [hunk 4]
```diff
@@ -313,17 +312,17 @@ async function formatFiles(context) {
     }
   }
 
-  for await (const pathOrError of expandPatterns(context)) {
-    if (typeof pathOrError === "object") {
-      context.logger.error(pathOrError.error);
+  for await (const { error, fileName, ignoreUnknown } of expandPatterns(
```
Comment (theodorejb) [inline src/cli/format.js:315]: Why rename `filename` to `fileName`? Filename is one word.

Comment (sosukesuzuki) [reply src/cli/format.js:315]: Because `expandPatterns` yields `fileName`. But I don't have any opinion on this naming.

Comment (fisker) [reply src/cli/format.js]: `filename` is fine when use alone, but when use together with `filePath` (I don't think `filepath` it's a word) and `fileUrl` , it's a little bit odd. So I use `fileName` sometime.

I'm not a native English speaker, not sure what's the best.

Comment (fisker) [reply src/cli/format.js]: And I use lots of `fileUrlOrPath` in recent PRs...

File: src/cli/expand-patterns.js [hunk 5]
```diff
@@ -20,15 +20,15 @@ async function* expandPatterns(context) {
       continue;
     }
 
-    const fileName = path.resolve(filePath);
+    const filename = path.resolve(filePath);
```
Comment (fisker) [inline src/cli/expand-patterns.js:23]: FYI: I'm waiting for fast-glob v4 to use `absolute: true`.  There are bugs in v3. https://github.com/mrmlnc/fast-glob/issues/371

**Comment by sosukesuzuki** [comment]
@fisker What do you think about release this as a patch version? We should release this a minor?

**Comment by fisker** [comment]
Minor should be better, since it's not a bug fix, we don't format them in v2 before, this become a problem only because we removed the plugin search feature.

We don't add much  new features recently.  Maybe patch is fine too. I'll leave the decision to you.

**Comment by sosukesuzuki** [comment]
I'll release this as a minor (`3.1`)

**Comment by t-ricci-digit** [comment]
Waiting for 3.1 release, to update my project dependencies

**Comment by benmccann** [comment]
Is there any ETA for the 3.1 release? We've been keeping the Svelte project templates on v2, but users have upgrading to v3 on their own and keep hitting this. I'd love to see this fix released

**Comment by julienw** [comment]
Something's not clear to me: does that make prettier try to process files that have supported plugins, even if these plugins are not loaded?

Thanks

**Comment by fisker** [comment]
Yes, it will process every file but ignore files can't find a parser.

**Comment by julienw** [comment]
> Yes, it will process every file but ignore files can't find a parser.

We had to change our config file from https://github.com/firefox-devtools/profiler/blob/2c0cbe6023ac94e16df11748946c1cdbdb4f6e00%5E/.prettierrc.js
to 
https://github.com/firefox-devtools/profiler/blob/2c0cbe6023ac94e16df11748946c1cdbdb4f6e00/.prettierrc.js
That is: change `files: 'bin/*'` to `files: 'bin/*.js'`.
IMO this makes sense though: we were configuring prettier to use "espree" on all the files in this directory, but some files weren't javascript. The previous prettier version wouldn't parse the .sh files despite this configuration, while the new version does. I think the new behavior is more correct, however this is still a breaking change.

**Comment by fisker** [comment]
Sorry about that. Glad you figured how to make it work 👍

**Pull request merged**